src/mbgl/renderer/tile_pyramid.cpp

```diff
@@ -231,7 +233,14 @@ void TilePyramid::update(const std::vector<Immutable<style::LayerProperties>>& l
                     tilesIt->second->setNecessity(TileNecessity::Optional);
                     cache.add(tilesIt->first, std::move(tilesIt->second));
                 }
-                tiles.erase(tilesIt++);
+
+                if (tilesIt->second) {
+                    tilesIt->second->cancel();
+                    Scheduler::GetBackground()->schedule(
+                        [tile_{std::shared_ptr<Tile>(std::move(tiles.extract((tilesIt++)->first).mapped()))}]() {});
```

Comment:
Somehow my brain can't parse this capture group.

Reply:
I had to look up how to do this actually, haha. It extracts a value from the container and move it to a shared pointer inside the lambda's capture group. Doing it this way means we can keep using the iterator for the rest of the container.

Reply:
I don't follow, why would it invalidate the iterator?

I somehow never noticed this, but I think it has the same temporary lambda copy retention issue as in `TileCache`.

Reply:
Is `tiles.extract((tilesIt++)->first)` important here, vs. `tiles.extract(tilesIt++)`?  It's doing an extra key search.

Reply:
From when I worked on it, I believe `extract().mapped()` was my best option to prevent a copy of the element while also removing it from the container. Other options didn't look like they'd work.

It may have that lambda issue, I never noticed the other instance in `TileCache`. I also haven't seen the destructor running on the main thread. To be safe, I'd say do the same here as in `TileCache`.

Reply:
It seems to be called somewhat rarely.  I couldn't trigger it from the benchmark or interactive, but it happens a few times in the render tests, apparently related to GeoJSON, so it might show up in performance in some scenario related to that.

Reply:
> Is `tiles.extract((tilesIt++)->first)` important here, vs. `tiles.extract(tilesIt++)`? It's doing an extra key search.

If it compiles and works correctly then no, I suppose it isn't.

Reply:
>It seems to be called somewhat rarely.

If it gets called at all then I say we should fix it, there is always a possibility of threading issues if that is allowed to happen.